pkgs/by-name/cu/cups-brother-dcpt310/package.nix

```diff
@@ -0,0 +1,118 @@
+{ stdenv
+, stdenvNoCC
+, lib
+, fetchurl
+, perl
+, gnused
+, dpkg
+, makeWrapper
+, autoPatchelfHook
+, libredirect
+, gnugrep
+, coreutils
+, ghostscript
+, file
+, pkgsi686Linux
+}:
+
+stdenvNoCC.mkDerivation rec {
+  pname = "cups-brother-dcpt310";
+  version = "1.0.1";
+  src = fetchurl {
+    url = "https://download.brother.com/welcome/dlf103618/dcpt310pdrv-${version}-0.i386.deb";
+    sha256 = "0g9hylmpgmzd6k9lxjy32c7pxbzj6gr9sfaahxj3xzqyar05amdx";
+  };
+
+  nativeBuildInputs = [ dpkg makeWrapper autoPatchelfHook ];
+
+  buildInputs = [ perl gnused libredirect pkgsi686Linux.stdenv.cc.cc.lib ];
+
+  unpackPhase = "dpkg-deb -x $src .";
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out"
+    cp -pr opt "$out"
+    cp -pr usr/bin "$out/bin"
+    rm "$out/opt/brother/Printers/dcpt310/cupswrapper/cupswrapperdcpt310"
+
+    mkdir -p "$out/lib/cups/filter" "$out/share/cups/model"
+
+    ln -s "../../../opt/brother/Printers/dcpt310/cupswrapper/brother_lpdwrapper_dcpt310" \
+      "$out/lib/cups/filter/brother_lpdwrapper_dcpt310"
+    ln -s "../../../opt/brother/Printers/dcpt310/cupswrapper/brother_dcpt310_printer_en.ppd" \
+      "$out/share/cups/model/brother_dcpt310_printer_en.ppd"
+
+    runHook postInstall
+  '';
+
+  # Fix global references and replace auto discovery mechanism
+  # with hardcoded values.
+  #
+  # The configuration binary 'brprintconf_dcpt310' and lpd filter
+  # 'brdcpt310filter' has hardcoded /opt format strings.  There isn't
+  # sufficient space in the binaries to substitute a path in the store, so use
+  # libredirect to get it to see the correct path.  The configuration binary
+  # also uses this format string to print configuration locations.  Here the
+  # wrapper output is processed to point into the correct location in the
+  # store.
+
+  postFixup = ''
+    interpreter=${pkgsi686Linux.glibc.out}/lib/ld-linux.so.2
+
+    substituteInPlace $out/opt/brother/Printers/dcpt310/lpd/filter_dcpt310 \
+      --replace "my \$BR_PRT_PATH =" "my \$BR_PRT_PATH = \"$out/opt/brother/Printers/dcpt310/\"; #" \
+      --replace /usr/bin/pdf2ps "${ghostscript}/bin/pdf2ps" \
+      --replace "my \$GHOST_SCRIPT" "my \$GHOST_SCRIPT = \"${ghostscript}/bin/gs\"; #" \
+      --replace "PRINTER =~" "PRINTER = \"dcpt310\"; #"
+
+    substituteInPlace $out/opt/brother/Printers/dcpt310/cupswrapper/brother_lpdwrapper_dcpt310 \
+      --replace "PRINTER =~" "PRINTER = \"dcpt310\"; #" \
+      --replace "my \$basedir = \`readlink \$0\`" "my \$basedir = \"$out/opt/brother/Printers/dcpt310/\"" \
+      --replace "my \$lpdconf = \$LPDCONFIGEXE.\$PRINTER;" "my \$lpdconf = \"$out/bin/brprintconf_dcpt310\";"
+
+
+    patchelf --set-interpreter "$interpreter" "$out/opt/brother/Printers/dcpt310/lpd/brdcpt310filter" \
+      --set-rpath ${ lib.makeLibraryPath [ pkgsi686Linux.stdenv.cc.cc ] }
```

Comment:
Shouldn't 64-bit `cc` be used on 64-bit systems? I know that they're backwards compatible with their 32-bit versions, but I'd rather use the correct binaries.

Reply:
This is for setting the library path of a 32-bit executable (distributed from Brother already as a binary). A 32-bit executable cannot use 64-bit libraries.

The rpath is set becauce the program needs `libstdc++.so.6` and `libgcc_s.so.1` findable within the library path. Maybe it is strange to use stdenv.cc.cc instead of directly using a gcc related package for this provided binary, but it is [common historically in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20makelibrarypath%20stdenv.cc.cc&type=code) for whatever reason. We could change `pkgsi686Linux.stdenv.cc.cc` to `pkgsi686Linux.gcc.cc` which looks to be more correct if this package is in scope of a user [override of stdenv's gcc to clang](https://nixos.wiki/wiki/Using_Clang_instead_of_GCC).